### PR TITLE
O(n^3) maximum weighted matching

### PR DESCRIPTION
**Pull request by yi-ji**
This pull request contains:

- `include/boost/graph/maximum_weighted_matching.hpp`
 An O(|V|^3) implementation of maximum weighted matching algorithm for general graphs, and a verifier using brute-force searching.

- `test/weighted_matching_test.cpp`
Tests on randomly generated graphs using the brute-force verifier.

- `example/weighted_matching_example.cpp`
Some simple usage examples.

The O(|V|^3) implementation refers to:
Eugene L. Lawler, _Combinatorial Optimization: Networks and Matroids_, Chapter 6.10
(Note that there are a few mistakes in the book.)

Thanks to @aaw for his good implementation of maximum cardinality matching that helped me clear my mind when starting.

### CODE REVIEW

**Comment by anadon** [comment]
I'm helping out with the PR backlog. Looks like you have a feature, test, and a new example. This will be in the second batch of merged PR's. This is to let you know and help me prioritize PR's.

**Comment by yi-ji** [comment]
@anadon Great news, really thank you for what you are doing!
Please let me know if you need anything :)

**Comment by anadon** [comment]
@yi-ji Please rebase your PR on the current devel.  There are some CI tools which have been enabled and a number of important cleanups.

File: test/weighted_matching_test.cpp [hunk 1]
```diff
@@ -108,6 +108,8 @@ void weighted_matching_test(typename graph_traits<Graph>::vertices_size_type num
                 ++num_edges;
         }
     }
+
+    //print_graph(g);
```
Comment (anadon) [inline test/weighted_matching_test.cpp]: Should this be removed?

File: test/weighted_matching_test.cpp [hunk 2]
```diff
@@ -147,8 +149,8 @@ int test_main(int, char*[])
     // you can also lower the max_num_v and/or max_num_e
     
     std::size_t max_num_v = 16;
-    std::size_t max_num_e = 20;
-    std::size_t batch_size = 100;
+    std::size_t max_num_e = 24;
+    std::size_t batch_size = 16;
```
Comment (anadon) [inline test/weighted_matching_test.cpp]: How do you calculate the optimal test size?

File: include/boost/graph/maximum_weighted_matching.hpp [hunk 3]
```diff
@@ -54,7 +54,7 @@ namespace boost
         {
             typedef boost::iterator_property_map<typename std::vector<T>::iterator, VertexIndexMap> type;
         };
-        typedef decltype(graph::detail::V_EVEN) vertex_state_t; // anonymous enum defined elsewhere
+        typedef unsigned short vertex_state_t;
```
Comment (anadon) [inline include/boost/graph/maximum_weighted_matching.hpp]: Does this type ever change in graph?

File: test/weighted_matching_test.cpp (hunk 1 above)
Comment (yi-ji) [reply test/weighted_matching_test.cpp]: Yes, sorry didn't notice, will do

File: include/boost/graph/maximum_weighted_matching.hpp (hunk 3 above)
Comment (yi-ji) [reply include/boost/graph/maximum_weighted_matching.hpp]: It's defined as anonymous enum and used only in `graph/max_cardinality_matching.hpp`,
see: https://github.com/boostorg/graph/blob/develop/include/boost/graph/max_cardinality_matching.hpp#L29

Comment (anadon) [reply include/boost/graph/maximum_weighted_matching.hpp]: Then could it be removed entirely and not just its use here?

File: test/weighted_matching_test.cpp (hunk 2 above)
Comment (yi-ji) [reply test/weighted_matching_test.cpp]: I didn't, because I am using a brute-force verifier, it can be time-consuming when edge num increases (exponential), and travis-ci has time limit

Comment (anadon) [reply test/weighted_matching_test.cpp]: Why not optimize it locally first?

Comment (yi-ji) [reply test/weighted_matching_test.cpp]: I can wait for as long as it takes on my machine, but I didn't know about travis-ci time limit

Comment (anadon) [reply test/weighted_matching_test.cpp]: I'm sure we can increase the time limit if it increases correctness.  But you should be able to determine the rough upper and lower bounds of these variables for testing correctness.

File: include/boost/graph/maximum_weighted_matching.hpp (hunk 3 above)
Comment (yi-ji) [reply include/boost/graph/maximum_weighted_matching.hpp]: "Here" is `graph/maximum_weighted_matching.hpp`, this enum is defined (and used) in `graph/max_cardinality_matching.hpp`, not by me. 
Another idea is to assign a name for this enum in `graph/max_cardinality_matching.hpp` but I think maybe it's better not to tamper with existed files

Comment (anadon) [reply include/boost/graph/maximum_weighted_matching.hpp]: I'm a little uncomfortable with changing something which seems to be related to a template parameter.  If this is just for testing time, it is probably best to leave it be.

File: test/weighted_matching_test.cpp (hunk 2 above)
Comment (yi-ji) [reply test/weighted_matching_test.cpp]: You're right, I will give a rough time estimate regarding edge num

File: include/boost/graph/maximum_weighted_matching.hpp (hunk 3 above)
Comment (yi-ji) [reply include/boost/graph/maximum_weighted_matching.hpp]: Did you mean leave `decltype(graph::detail::V_EVEN)` be? it won't compile by g++
I understand you concern, however `vertex_state_t` has only possible values of {0, 1, 2} (no chance to be changed by logic) and only used in `graph/max_cardinality_matching.hpp`, so unsigned short shouldn't cause any problem I think

File: test/weighted_matching_test.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+#include <boost/graph/maximum_weighted_matching.hpp>
+
+#include <iostream>                      // for std::cout
+#include <boost/property_map/vector_property_map.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/adjacency_matrix.hpp>
+#include <boost/graph/random.hpp>
+#include <ctime>
+#include <boost/random.hpp>
+#include <boost/test/minimal.hpp>
```
Comment (jeremy-murphy) [inline test/weighted_matching_test.cpp]: I strongly recommend that you don't use randomly generated test cases here. Using them as a separate utility to discover bugs is up to you, but the test cases here should attempt to provide good coverage of the algorithm by providing inputs that execute different code paths, edge cases, etc.
Then you also don't need to run a brute-force verification: just store the answer and compare.

Comment (yi-ji) [reply test/weighted_matching_test.cpp]: Nice advice, thank you. I will remove brute-force verification and use pre-computed answers instead.

File: include/boost/graph/maximum_weighted_matching.hpp (hunk 3 above)
Comment (anadon) [reply include/boost/graph/maximum_weighted_matching.hpp]: OK.  This is new code so good enough for me at this point.  I'll read through the PR in more detail when the tests pass.

**Comment by anadon** [comment]
It looks like the anonymous enum to short is causing some other knock off effects causing compilation failure.

**Comment by yi-ji** [comment]
@anadon Not sure why jobs 87.17 and 87.18 failed because no output shown...seems related to clang++5.0? Though I tested locally without seeing anything wrong
Or might need help from this: https://docs.travis-ci.com/user/common-build-problems/#ruby-tests-frozen-and-cancelled-after-10-minute-log-silence

**Comment by anadon** [comment]
@jzmaddock This looks good to me to merge.  It provides an interesting addition to BGL.

**Comment by anadon** [comment]
@yi-ji Make sure to add documentation and examples.

**Comment by jzmaddock** [comment]
We have an example, but we do need docs before merging.  Please add "[CI SKIP]" to pure documentation commit messages to stop CI builds cycling needlessly: we're short on CI testing time especially on appveyor which seems to all run through my account :(

**Comment by anadon** [comment]
Would it help if I set up something through appveyor to run these through?

**Comment by yi-ji** [comment]
@anadon @jzmaddock documentation added :)

**Comment by anadon** [comment]
I'm not seeing CI run.  Do you have an idea of what is up with that?

**Comment by yi-ji** [comment]
@anadon  I added "[CI SKIP]" into the commit message as @jzmaddock [asked](https://github.com/boostorg/graph/pull/110#issuecomment-435134621). No need to run a CI build for pure document commits, right?

**Comment by yi-ji** [comment]
@jzmaddock Could you help merge this PR?

**Comment by jeremy-murphy** [comment]
If you all don't mind, I would like to have a good look through the algorithm implementation before we merge. There is a lot of code there.
I see some use of `shared_ptr` and virtual functions that I'm a bit suspicious about.

And someone remind me... is Boost.Graph still on a strict C++03 diet?

**Comment by jzmaddock** [comment]
>If you all don't mind, I would like to have a good look through the algorithm implementation before we merge. There is a lot of code there.
>I see some use of shared_ptr and virtual functions that I'm a bit suspicious about.

It would be great if someone would give this a once over.

>And someone remind me... is Boost.Graph still on a strict C++03 diet?

I'd prefer to keep existing code passing with existing compilers.  But for strict additions, I don't see why we can't use C++20 if that's the only way to do it.  As long as it's documented of course.  In any case this looks all C++03 to me at first glance, and the few things that have C++11 equivalents (shared_ptr, tie) are internal details so far as I can see.

File: doc/bibliography.html [hunk 5]
```diff
@@ -438,6 +438,16 @@ <H2>Bibliography</H2>
...
+<p></p><dt><a name="edmonds65:_max_weighted_match">74</a>
+<dd>J. Edmonds<br>
+<em>Maximum Matching and a Polyhedron with 0, 1-Vertices</em><br>
+Journal of Research of the National Bureau of Standards B 69, pp. 125-130, 1965.
+
+<p></p><dt><a name="gabow90">75</a>
+<dd>Harold N. Gabow<br>
+<em>Data Structures for Weighted Matching and Nearest Common Ancestors with Linking</em><br>
+Proceedings of the First Annual ACM-SIAM Symposium on Discrete Algorithms, pp. 434-443, 1990.
+
```
Comment (jeremy-murphy) [inline doc/bibliography.html]: I know this publication from 1990 is more recent, but maybe his original paper on the implementation is more relevant and specific? https://dl.acm.org/citation.cfm?id=321942

File: doc/maximum_weighted_matching.html [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+<a name="sec:weighted_matching">Before you continue, it is recommended to read
+about <a href="./maximum_matching.html">maximal cardinality matching</a> first. 
+A <i>maximum weighted matching</i> of an edge-weighted graph is a matching
+for which the sum of the weights of the edges is maximum. 
+Two different matchings (edges in the matching are colored blue) in the same graph are illustrated below. 
+The matching on the left is a maximum cardinality matching of size 8 and a maximal
+weighted matching of weight sum 30, meaning that is has maximum size over all matchings in the graph
+and its weight sum can't be increased by adding edges.
+The matching on the right is a maximum weighted matching of size 7 and weight sum 38, meaning that it has maximum
+weight sum over all matchings in the graph.
```
Comment (jeremy-murphy) [inline doc/maximum_weighted_matching.html]: In terms of nomenclature, Gabow (1990) mentions a _maximum weight matching_ problem but in his earlier 1976 paper the problem is called _maximum matching_, as it is by Edmonds in his original paper.
My guess is that he added "weight" in to differentiate against the _maximum cardinality matching_ problem.
But either way, I'm going to suggest that it is "weight", not "weighted", although I know he uses the latter in some other names. Anyway, it's not a serious issue, just my pedantic opinion.

In addition to that, the 1990 paper raises another point, that this O(n^3) algorithm is not the only algorithm that solves maximum weight matching, as it presents an O(n(m + n log n)) algorithm. It's perfectly reasonable to implement the O(n^3) algorithm, but we need to name it such that we can also name the newer algorithm, e.g.:

- `brute_force_maximum_weight_matching` -- O(?)
- `label_encoding_maximum_weight_matching` -- O(n^3)
- `blossom_expansion_maximum_weight_matching` -- O(n(m + n log n))

They are just ideas, I'm not sure if they are entirely accurate, especially the last one.

Then, maybe, we could have a `maximum_weight_matching` that uses some heuristics to pick the fastest algorithm for a given problem size. Just an idea, not important.

File: doc/bibliography.html (hunk 5 above)
Comment (yi-ji) [reply doc/bibliography.html]: I agreed, change will be made. 
Just to be clear, the labeling method is not generalized to weighted matching in the paper you mentioned, though it does can be. 
I have to admit that the 1990 paper focused more about improving the time complexity from O(n^3) to O(n(m+nlogn)) using link-cut-trees to find nearest common ancestor, which is not implemented in this PR..

File: doc/maximum_weighted_matching.html (hunk 6 above)
Comment (yi-ji) [reply doc/maximum_weighted_matching.html]: I believe both "maximum weight / weighted matching" are commonly used. In fact, another large C++ template library for network algos, [LEMON](http://lemon.cs.elte.hu/pub/doc/latest/a00254.html), uses `MaxWeightedMatching`.

Thanks for the name prefix suggestion. However the examples might not be quite accurate. The O(n^3) and O(n(m + n log n)) algorithms both have labeling and blossom expansion operations. The only difference is the O(n(m + n log n)) algorithms uses link-cut-tree to find nearest common ansestors, which is not important/radical/obvious enough to call them with different names.
I didn't implement the O(n(m + n log n)) algo mainly because there is no available link-cut-tree module in boost. If it becomes available and someone wants to improve to O(n(m + n log n)) someday, he/she could just make a few lines of code change and doesn't have to keep an O(n^3) version.

File: doc/bibliography.html (hunk 5 above)
Comment (jeremy-murphy) [reply doc/bibliography.html]: I'm not quite sure what you mean about the labeling method is not generalized in that paper?

But otherwise, glad my suggestion was helpful.

I didn't realize that NCA (which I think is typically called LCA?) is part of this algorithm, although I haven't finished going through it yet. Where is it in here?

File: doc/maximum_weighted_matching.html (hunk 6 above)
Comment (jeremy-murphy) [reply doc/maximum_weighted_matching.html]: Ah, hmmm, OK, you think the difference between the O(n^3) and O(n(m + n log n)) algorithms will be a matter of passing in different LCA sub-algorithms? That would certainly make the naming easier.

Funnily enough, I am actually working on... close to finishing... a pragmatic implementation of LCA, but it's not Gabow's  link-cut version.  When I have a chance, I'll have a look at how much more work it would be to extend it, although I want to finish the basic implementation first.

File: doc/bibliography.html (hunk 5 above)
Comment (yi-ji) [reply doc/bibliography.html]: I mean, the labeling procedure in the paper you mentioned only applies to unweighted matching. The author just claimed at the end that it could be generalized to weighted matching.

Link-cut-tree based NCA/LCA is not implemented in this PR. I am using a trivial method, which is used in `max_cardinality_matching.hpp` as well.

In fact, this 1990 citation should not be deleted... I didn’t look carefully where it is cited when replying you. It’s not cited for algorithm reference, but for pointing out the best known time complexity is not O(n^3).
I cited Edmond: “Maximum Matching and a Polyhedron with 0, 1-Vertices” for my implementation. 
Of course I will still add the paper you mentioned, because Edmond’s paper did not elaborate implementation detail for labeling procedure.

File: doc/maximum_weighted_matching.html (hunk 6 above)
Comment (yi-ji) [reply doc/maximum_weighted_matching.html]: Thanks. More funnily... I am also working on and close to finishing a link-cut-tree module😂 but I believe it’s okay if we open dIfferent PR. Let’s see what to do then.

**Comment by yi-ji** [comment]
Review comments resolved. Ready for merge again! (If no more of them)

**Comment by jeremy-murphy** [comment]
I haven't had a chance to look at the code itself in detail yet and realistically I won't until the new year. I'm inclined to say that so long as it is well tested (which I think it is), let's not let perfect be the enemy of good, merge it now, and look at refactoring when we have time.

**Comment by yi-ji** [comment]
Happy 2019 guys.. so is there anything stalling the merge...?

**Comment by jzmaddock** [comment]
Merged.
I've re-ordered the initialization of members of maximum_weighted_matching to avoid pages and pages of gcc warnings.  Also added an entry to the documentation table-of-contents.
